src/codegen/compile.ml

```diff
@@ -3411,6 +3417,9 @@ module Arr = struct
       idx env
   )
 
+  let element_type env typ = match typ with
+    | Type.Array (element_type) -> element_type
+    | _ -> assert false
```

Comment:
This worries me slightly if you are trying to access the type of the element T in e.g. [var T] or [T].
I wonder if it should be
```suggestion
  let element_type env typ = 
   Type.(match normalize typ with
    | Array (Mut element_type)  
    | Array element_typ -> element_type
    | _ -> assert false)
```

The other thing I'm not sure of is whether T could be a typ variable, in which case it might be necessary to promote (to its bound) instead of normalize.  


An example to test whether this works with code that requires promotion might be:

```
func sub<T, A <: [Var T]>(a : A) : T { a[0]; }
```

Reply:
Thank you very much.
There was indeed an issue with generic array type definitions that I now fixed by type promotion.
Handling mutable and non-mutable types is covered by `potential_pointer_type` and can be omitted in this function.